src/GraphTopology.cpp

```diff
@@ -13,469 +13,301 @@
-    Agg_vertex_.Swap(tmp);
+    auto starts = parlinalgcpp::GenerateOffsets(comm, {num_vertices, num_edges, num_aggs});
+    const auto& vertex_starts = starts[0];
+    const auto& edge_starts = starts[1];
+    const auto& agg_starts = starts[2];
 
-    mfem::SparseMatrix aggregate_edge(smoothg::Mult(Agg_vertex_,
-                                                    vertex_edge));
+    ParMatrix edge_agg_d(comm, edge_starts, agg_starts, std::move(edge_ext_agg));
+    ParMatrix agg_edge_d = edge_agg_d.Transpose();
 
-    // Need to sort the edge indices to prevent index problem in face_edge
-    aggregate_edge.SortColumnIndices();
+    ParMatrix edge_agg_ext = edge_edge.Mult(edge_agg_d);
+    ParMatrix agg_agg = agg_edge_d.Mult(edge_agg_ext);
 
-    AggregateEdge2AggregateEdgeInt(aggregate_edge, Agg_edge_);
-    mfem::SparseMatrix edge_aggregate(smoothg::Transpose(aggregate_edge));
+    agg_edge_ext = 1.0;
+    SparseMatrix face_agg_int = MakeFaceAggInt(agg_agg);
+    SparseMatrix face_edge_ext = face_agg_int.Mult(agg_edge_ext);
 
-    // block diagonal edge_aggregate and aggregate_edge
-    auto edge_aggregate_d = make_unique<mfem::HypreParMatrix>(
-                                comm_, edge_start_.Last(), aggregate_start_.Last(), edge_start_,
-                                aggregate_start_, &edge_aggregate);
-    auto aggregate_edge_d = make_unique<mfem::HypreParMatrix>(
-                                comm_, aggregate_start_.Last(), edge_start_.Last(),
-                                aggregate_start_, edge_start_, &aggregate_edge);
+    face_edge_local_ = MakeFaceEdge(agg_agg, edge_agg_ext,
+                                    agg_edge_ext, face_edge_ext);
 
-    unique_ptr<mfem::HypreParMatrix> d_td_d_edge_Agg(
-        ParMult(edge_d_td_d_.get(), edge_aggregate_d.get()) );
-    unique_ptr<mfem::HypreParMatrix> Agg_Agg(
-        ParMult(aggregate_edge_d.get(), d_td_d_edge_Agg.get()) );
+    face_agg_local_ = ExtendFaceAgg(agg_agg, face_agg_int);
+    agg_face_local_ = face_agg_local_.Transpose();
 
-    auto Agg_Agg_d = ((hypre_ParCSRMatrix*) *Agg_Agg)->diag;
-    auto Agg_Agg_o = ((hypre_ParCSRMatrix*) *Agg_Agg)->offd;
+    auto face_starts = parlinalgcpp::GenerateOffsets(comm, face_agg_local_.Rows());
 
-    // nfaces_int = number of faces interior to this processor
-    HYPRE_Int nfaces_int = Agg_Agg_d->num_nonzeros - Agg_Agg_d->num_rows;
-    assert( nfaces_int % 2 == 0 );
-    nfaces_int /= 2;
+    face_edge_ = ParMatrix(comm, face_starts, edge_starts, face_edge_local_);
+    ParMatrix edge_face = face_edge_.Transpose();
 
-    // nfaces_bdr = number of global boundary faces in this processor
-    int nfaces_bdr = 0;
-    mfem::SparseMatrix aggregate_boundaryattr;
-    if (edge_boundaryattr)
-    {
-        mfem::SparseMatrix tmp = smoothg::Mult(aggregate_edge, *edge_boundaryattr);
-        aggregate_boundaryattr.Swap(tmp);
+    face_face_ = parlinalgcpp::RAP(edge_edge, edge_face);
```

Comment:
I am thinking if it worths to make a function for triple matrix product `Matrix Mult(Matrix&, Matrix&, Matrix&)`. I dont know about the new code, but I remember there were several places in the old code where we are doing this kind of RAP, but since the specific assumption of RAP function, we end up doing the transpose step twice. I dont think this extra step of transpose affects the performance that much, but just feel like wasted.

Another way could be to make a function call `PAR` (sounds confusing though), which transpose the other side.

Reply:
Yes this would be a helpful function.  Calling it `Mult` would probably be most clear since that is exactly what it will be doing, without any transpose.